### ISSUE TITLE
reserved identifier violation

### ISSUE THREAD

**Issue by elfring** [dotnet/coreclr#270]
I would like to point out that identifiers like "[```__CAPARSER_H__```](https://github.com/dotnet/coreclr/blob/ef1e2ab328087c61a6878c1e84f4fc5d710aebce/src/inc/caparser.h#L15)" and "[```__GC_H```](https://github.com/dotnet/coreclr/blob/ef1e2ab328087c61a6878c1e84f4fc5d710aebce/src/gc/gc.h#L15)" [do not fit](https://www.securecoding.cert.org/confluence/display/cplusplus/DCL32-CPP.+Do+not+declare+or+define+a+reserved+identifier#DCL32-CPP.Donotdeclareordefineareservedidentifier-NoncompliantCodeExample%28HeaderGuard%29 "Do not use identifiers that are reserved for the compiler implementation.") to the expected naming convention of the C++ language standard.
Would you like to adjust your selection for unique names?

**Comment by jkotas** [comment]
Majority of the CoreCLR code is compiled against PAL headers that isolate from system C++ headers, and so the identifiers cannot collide by design.

It would nice to clean this up for parts of the tree that are likely to evolve into reusable components - GC is one that we know about so far.

@elfring  We will be happy to accept PR that cleans up the header guards in files under src\gc. Thanks!

**Comment by elfring** [comment]
Should the affected source files be treated in a way so that they should be independent from a specific implementation of a C++ compiler?

**Comment by jkotas** [comment]
We would like to make the GC portable by centralizing all dependencies on a specific execution engine, runtime and compiler in the GC environment (gcenv.h). See comment in https://github.com/dotnet/coreclr/blob/master/src/gc/sample/GCSample.cpp.

There is a number of issues to fix to get there. We prefer this type of work to be split into small incremental pull requests - e.g. fix one particular type of issue per PR.

Also, check the contribution guidelines for GC at https://github.com/dotnet/coreclr/wiki/Garbage-Collector-Contributions

cc: @Maoni0 @brianrob

**Comment by Maoni0** [comment]
What Jan said is fine by me. Thanks.

**Comment by MattWhilden** [comment]
@elfring Did you get a chance to take a stab at this?

Also, would you mind if this title was changed to something like "C++ identifiers under src\gc need cleanup for portability"? We can open other issues to track work in other components as they are identified.

**Comment by elfring** [comment]
@MattWhilden: No.

Which safer name pattern do you prefer here?
How do you think about to make [your include guards](https://en.wikipedia.org/wiki/Include_guard#Difficulties "Consistency for include guards") not only standard-compliant but also really unique by appending [a kind of UUID](https://en.wikipedia.org/wiki/Universally_unique_identifier "universally unique identifier")?

**Comment by MattWhilden** [comment]
I don't really have a strong preference. Perhaps @jkotas or @Maoni0 do.

**Comment by jkotas** [comment]
Appending GUIDs makes the code look ugly and it is not really solving much. The remaining 99.9% of names can still collide. I do not think it is something we want to do.

Tweaking the number of underscores to make the guards more compliant with these guidelines is fine with me. Alternatively, changing them to #pragma once is ok too - though it is non-standard, but all current compilers support it.

**Comment by elfring** [comment]
Standard-compliant identifiers do not need to win a beauty competition, do they?

**Comment by Maoni0** [comment]
I agree with Jan - either #pragma once or getting rid of the double underscore is fine.
